### DISCUSSION TITLE
imgsz issue - huge performance difference if I downscale before or during inference

### DISCUSSION THREAD

**Discussion by mfkeles** [ultralytics/ultralytics#21702] [Q&A]
**Body:**
I’ve noticed a big performance gap when I downscale videos before running YOLO inference, depending on whether I use FFmpeg or OpenCV.

**Setup:**

* **Training data:** 550×400 frames (downscaled from original 1100×800 source videos).
* **Training command:**

```python
# Train the model with 2 GPUs
results = model.train(
    data=os.path.join(working_path, 'data.yaml'), 
    epochs=200, 
    imgsz=640,  # I now know default is already 640
    device=[0, 1],
    batch=64,
    project=os.path.join(working_path, 'runs'),
    name='exp_yolo11l_200epoch'
)
```

* Original source videos are 1100×800.
* Model performs **very poorly** when I run inference on the original 1100×800 videos.
* If I **downscale to 550×400 with FFmpeg**, predictions are **great**.
* If I try to resize to 550×400 with OpenCV, results are **similar to the poor original-size results**, not the good FFmpeg results.

**FFmpeg downscale command:**

```bash
ffmpeg -i input.mp4 -vf scale=iw/2:ih/2 -c:a copy output.mp4
```

**OpenCV downscale:**

```python
frame_half = cv2.resize(frame, (w // 2, h // 2), interpolation=cv2.INTER_AREA)
```

**Batch inference loop:**

```python
def run_inference(frames, indices):
    if not frames:
        return
    preds = model(frames, imgsz=640, conf=conf, verbose=False)  # No imgsz override
    _collect_preds(preds, indices, results_list)

while True:
    ret, frame = cap.read()
    if not ret:
        break
    batch_frames.append(frame)
    batch_indices.append(frame_idx)
    frame_idx += 1
    if len(batch_frames) == batch_size:
        run_inference(batch_frames, batch_indices)
        if pbar: pbar.update(len(batch_frames))
        batch_frames, batch_indices = [], []
if batch_frames:
    run_inference(batch_frames, batch_indices)
    if pbar: pbar.update(len(batch_frames))
```

**What I’ve observed:**

* Feeding 1100×800 directly → poor predictions.
* FFmpeg-downscaled to 550×400 → great predictions.
* OpenCV-downscaled to 550×400 → poor predictions, similar to feeding 1100×800 directly.
* Almost identical results between feeding full res vs resizing with OpenCV
* Both methods produce videos of the same nominal resolution, but outputs from FFmpeg clearly work better with my model.

**Questions:**

1. How exactly does YOLO handle `imgsz` and internal letterboxing when the input is already pre-scaled?
2. Could FFmpeg’s scaling pipeline (filters, color space handling, padding, or alignment) be producing something fundamentally different that interacts better with YOLO than OpenCV’s output?
3. Has anyone else seen this gap? If so, what settings or preprocessing steps make OpenCV resizing match FFmpeg’s results?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @mfkeles, thanks for starting the discussion and for the detailed context 🚀 This is an automated response to help you get quicker traction. An Ultralytics engineer will also follow up here soon.

We recommend a visit to the Docs where you can find many Python and CLI usage examples and where many of the most common questions may already be answered:
- Docs: https://docs.ultralytics.com/
- Python examples: https://docs.ultralytics.com/usage/python/
- CLI examples: https://docs.ultralytics.com/usage/cli/

If this is a 🐛 Bug Report, please provide a minimum reproducible example to help us debug it:
- MRE guide: https://docs.ultralytics.com/help/minimum-reproducible-example/

If this is a custom training ❓ Question, please include:
- Dataset samples showing the issue
- Exact preprocessing steps (FFmpeg and OpenCV commands)
- Full training/inference logs and environment details
- Verify you’re following our Tips for Best Training Results: https://docs.ultralytics.com/guides/model-training-tips/

Join the Ultralytics community where it suits you best:
- Discord (real-time chat) 🎧: https://discord.com/invite/ultralytics
- Discourse (in-depth Q&A): https://community.ultralytics.com/
- Subreddit (community threads): https://www.reddit.com/r/Ultralytics/

## Upgrade
Please upgrade to the latest ultralytics package including all requirements in a Python>=3.8 environment with PyTorch>=1.8 to verify the behavior on the latest version:
```bash
pip install -U ultralytics
```
- Requirements: https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml

## Environments
YOLO may be run in any of the following up-to-date verified environments (with all dependencies including CUDA/CUDNN, Python and PyTorch preinstalled):
- Notebooks with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- Google Cloud Deep Learning VM. See GCP Quickstart Guide: https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/
- Amazon Deep Learning AMI. See AWS Quickstart Guide: https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/
- Docker Image. See Docker Quickstart Guide: https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/ <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>
If this badge is green, all Ultralytics CI tests are currently passing. CI tests verify correct operation of all YOLO Modes and Tasks on macOS, Windows, and Ubuntu every 24 hours and on every commit.

To help us assist you faster regarding your imgsz and preprocessing observations, please share:
- A minimal script that reproduces the difference between FFmpeg vs OpenCV resizing using your loop (MRE)
- Exact FFmpeg command and OpenCV code used (including any color conversion, interpolation, and letterboxing)
- A few representative frames before/after each preprocessing path
- The exact inference call and parameters you’re using (imgsz, conf, device, etc.)
- Your ultralytics version, OS, Python, PyTorch, CUDA, and GPU details

Thank you for the thorough breakdown—once we have the MRE and environment details, we’ll dig in further 💡

**Comment by Y-T-G** [comment]
What downscaling method was used for the training images?

**Comment by mfkeles** [comment]
I extracted images after downscaling with ffmpeg.

**Reply by Y-T-G** [reply]
Since you trained with that, the model performs better on that downscaling method

**Comment by yyq19990828** [comment]
It's normal. Models(cnn-based) can only adapt to similar scales you used for training when infering.

**Reply by glenn-jocher** [reply]
Good find—this is expected since you trained on FFmpeg-resized frames, so the model has adapted to that scaler/color pipeline. For predict, YOLO always resizes to imgsz with aspect-ratio letterboxing to stride; if you pre-resize then still pass imgsz=640, you’re effectively resampling twice. To align results: either keep FFmpeg-resized inputs and set imgsz to the exact frame size (tuple h, w) so no extra resize occurs, or match FFmpeg in OpenCV (use a bicubic/Lanczos kernel and consistent color matrix). For robustness, retrain using the same preprocessing you’ll use at inference. Details on imgsz/rect are in the Model Prediction inference arguments.
